idc_index/index.py

```diff
@@ -502,6 +465,364 @@ def get_viewer_URL(
+        Lastly, updates the manifest from generic download directory to explicitly mentioned directory
+        Args:
+            manifestFile (str): The path to the manifest file.
+            downloadDir (str): The path to the download directory.
+        Returns:
+            total_size (float): The total size of all series in the manifest file.
+            endpoint_to_use (str): The endpoint URL to use (either AWS or GCP).
+            temp_manifest_file(Path): Path to the temporary manifest file for downstream steps
+        Raises:
+            ValueError: If the manifest file does not exist, if any URL in the manifest file is invalid, or if any URL is inaccessible in both AWS and GCP.
+            Exception: If the manifest contains URLs from both AWS and GCP.
+        """
+
+        # Read the csv file
+        manifest_df = pd.read_csv(
+            manifestFile, comment="#", skip_blank_lines=True, header=None
+        )
+
+        # Rename the column
+        manifest_df.columns = ["manifest_cp_cmd"]
+
+        # Apply the function to the 'manifest_cp_cmd' column
+        manifest_df["manifest_crdc_instance_uuid"] = manifest_df[
+            "manifest_cp_cmd"
+        ].apply(self._extract_crdc_instance_uuid_from_manifest_s3_url)
+
+        # Remove 'cp ' at the beginning and ' .' at the end of the 'manifest_cp_cmd'
+        manifest_df["s3_url"] = (
+            manifest_df["manifest_cp_cmd"]
+            .str.replace(r"^cp ", "", regex=True)
+            .str.replace(r" \.$", "", regex=True)
+        )
+        index_df_copy = self.index
+        index_df_copy["index_crdc_instance_uuid"] = index_df_copy[
+            "series_aws_url"
+        ].apply(self._extract_crdc_instance_uuid_from_series_aws_url)
+        # Perform left join on client.index using crdc_instance_uuid
+        merged_df = manifest_df.merge(
+            index_df_copy,
+            left_on="manifest_crdc_instance_uuid",
+            right_on="index_crdc_instance_uuid",
+            how="left",
+        )
+
+        # Create 'crdc_instance_uuid_found' column
+        merged_df["crdc_instance_uuid_found"] = merged_df[
+            "index_crdc_instance_uuid"
+        ].notna()
+
+        # Check if crdc_instance_uuid is found in the index
+        if not all(merged_df["crdc_instance_uuid_found"]):
+            missing_manifest_cp_cmds = merged_df.loc[
+                ~merged_df["crdc_instance_uuid_found"], "manifest_cp_cmd"
+            ]
+            missing_manifest_cp_cmds_str = f"The following manifest copy commands do not have any associated series in the index: {missing_manifest_cp_cmds.tolist()}"
+            raise ValueError(missing_manifest_cp_cmds_str)
+
+        # Create 'endpoint' column
+        merged_df["endpoint"] = merged_df.apply(
+            lambda row: aws_endpoint_url
+            if row["crdc_instance_uuid_found"]
+            and row["s3_url"] == row["series_aws_url"]
+            else gcp_endpoint_url,
```

Comment:
I do not want to make this assumption.

I suggest for now we simplify the logic of the endpoint identification, and keep it the same as before: check the first URL, whatever it is - AWS or GCP - assume it is the same for all URLs and proceed with the download using that endpoint.

If the URL is invalid, I hope it is possible to communicate to the user which specific URL failed when copy is aborted.

We can have an option in the function to perform validation of the manifest, to do more thorough checks, but I think in most cases those are not going to be useful, since manifest will either come from us, or from the portal.

What do you think?

Reply:
In the main branch implementation, the manifest is generalized by just looking at the first line. There could be invalid urls from second line and no

https://github.com/ImagingDataCommons/idc-index/blob/main/idc_index/index.py#L621

Compared to that, I'd argue that ths PR implementation is more rubust, not very time expensive (~3 seconds to validate 65k series manifest). But I do realize that it is not perfect and drew the line after checking a valid crdc uuid.

It is a battle of two assumptions. Checking the first line and no error handling vs checking every line for a valid crdc uuid and raising an exception if not. If you think, we can assume that the manifest is valid by looking at just the first line, I'm fine with that and revert the changes.

Reply:
My concern is that in the code highlighted, if we do not find a match, we cannot really conclude that the bucket is in GCP, and that's what the code is doing right now. I do not object to matching UUIDs (although we should check how long it takes for a 400K manifest).

It is not quite a battle of two assumption. In either case, if UUID/bucket are invalid, the download operation will abort. The difference is that currently it will abort with a message "The manifest contains URLs from both AWS and GCP. Please use only one provider", which may be rather misleading. If the UUID/bucket path is invalid, it will fail to find the file and I assume will print the corresponding message.

Reply:
> "The manifest contains URLs from both AWS and GCP. Please use only one provider", which may be rather misleading. If the UUID/bucket path is invalid, it will fail to find the file and I assume will print the corresponding message.

This is not true completely. Yes, it is possible for this message to be incorrect but only when crdc uuid is **FOUND** but the bucket ID is invalid.

Please see this code which is run before the cited code.

```

# Check if crdc_instance_uuid is found in the index
if not all(merged_df["crdc_instance_uuid_found"]):
    missing_manifest_cp_cmds = merged_df.loc[
        ~merged_df["crdc_instance_uuid_found"], "manifest_cp_cmd"
    ]
    missing_manifest_cp_cmds_str = f"The following manifest copy commands do not have any associated series in the index: {missing_manifest_cp_cmds.tolist()}"
    raise ValueError(missing_manifest_cp_cmds_str)
```


 I';l try another visual way to explain this.

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/vamsi/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/vamsi/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"Aptos Narrow", sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->

</head>

<body link="#467886" vlink="#96607D">


  |   |   | End point to use
-- | -- | -- | --
  | crdc uuid match | complete s3 url match |  
Case 1 | TRUE | TRUE | AWS
Case 2 | TRUE | FALSE | GCS
Case 3 | FALSE | N/A | Raise an exception
Case 4 | TRUE | Mix of TRUE and FALSE | Raise an exception that manifest has both GCS and AWS endpoints



</body>

</html>


> In either case, if UUID/bucket are invalid, the download operation will abort.

In the main branch's implementation, download begins even with an invalid manifest, i.e until it encounters an invalid line. This PR does not begin download until the manifest is valid for aws manifests. For GCP manifests, yes, it will also download until it encounters an invalid bucket path. 

Could I know what you would like me to do, after seeing the pros and cons of the PR implementation? 

Revert to checking only the first url only or keep the PR implementation?

Reply:
I think I understand what you are doing. My initial comment here was about the assumption that an unknown URL is a GCP URL, and the error message that stems from that assumption.

I suggest we keep the UUID check, and use the first URL check whenever we do not have a match with AWS URL to determine if the endpoint is GCP.

Reply:
I thought about this a bit, and I think we can improve this a lot with a minimal effort.

We can add this query and the result to `idc-index-data`, and it will give us mapping between AWS and GCP buckets:

```sql
SELECT
  DISTINCT(aws_bucket),
  gcs_bucket
FROM
  `bigquery-public-data.idc_current.dicom_all`
```

Then we can maintain this table separately and use it for validating GCS URLs and also for providing GCS URLs if requested by user.

It is fine to do this later. Let's just wrap this PR first.

Reply:
> I suggest we keep the UUID check, and use the first URL check whenever we do not have a match with AWS URL to determine if the endpoint is GCP.

Sounds good to me.


>  thought about this a bit, and I think we can improve this a lot with a minimal effort.

> We can add this query and the result to idc-index-data, and it will give us mapping between AWS and GCP buckets:

I know of a solution but you may not like it. In my opinion, there is no reason to have duplicate content in any columns. i.e we must create nesting. As I have mentioned before, users will never have to unnest. It is us, who have the expertise to unnest could take advantage, to sacrifice a little bit of time for sustainability. But that's a rant for another day.

Reply:
> I know of a solution but you may not like it. In my opinion, there is no reason to have duplicate content in any columns. i.e we must create nesting.

pip package size is a precious and limited resource. Why would we want to keep bucket information in any of the columns of the main index - nested or not - if we can do static mapping with just 3 rows? Again, I do not understand the motivation.

Reply:
Exactly my point as well that pip size is precious. But I meant in general. Collection_id, patientid, study id  and several others columns have only a fraction of unique values and there is absolutely no need to populate them in every line when we can nest anything common above series level attributes. 

I'll write a nested query and show you the savings from nesting vs the current one. But yes it will be populated in more than 3 rows for sure. 

Essentially my point is equivalent to splitting our one flatenned table to several small tables and joining just like the way you are referring to the bucket mapping.

Reply:
I agree we should consider different alternatives. Let's address this outside of this PR.

Reply:
I suggest we do the following:

1. make this check optional as well
2. keep this check enabled by default
3. add console log message to the user informing about what is going on (ie, "Checking validity of manifest UUIDs ...").

Reply:
this is an internal function, we can introduce a new parameter called `validate_manifest` in `download_from_manifest` and call this internal function only when `validate_manifest`=True. Otherwise, the manifest is passed directly to _s5cmd_run.

I'll add the logging message.

Reply:
> Otherwise, the manifest is passed directly to _s5cmd_run.

I realize it's more complicated than that. We need to parse each line regardless to get total size. I'll still pass the manifest to validator but ignore all errors when `validate_manifest= False`